(#170) Refactor cc_trust_data() constructor to fix bug.

gapisback opened the pull request:
Default constructor for cc_trust_data() ran into seg-faults due to use of uninitialized variables. Refactor the constructor to add  a common private `cc_trust_data_default_init()` method, shared by both constructors. 

Add Python tests which trigger this bug and test the fix.

Good catch!

The only concern I have is that the destructor does not delete the certifiers structures that were newed when adding.

You are correct that the code that does that, below, should be in the destructor (Man, that was a wierd mistake).

  for (int i = 0; i < num_certified_domains_; i++) {
      if (certified_domains_[i] != nullptr) {
        delete certified_domains_[i];
        certified_domains_[i] = nullptr;
      }
  }
  delete certified_domains_;
  certified_domains_ = nullptr;
  num_certified_domains_ = 0;

gapisback commented on src/cc_helpers.cc:
```diff
@@ -82,16 +82,22 @@ extern string gramine_platform_cert;
 certifier::framework::cc_trust_data::cc_trust_data(const string& enclave_type, const string& purpose,
     const string& policy_store_name) {
-
+  cc_trust_data_default_init();
```
Rearrange the constructor to do the "default initialization" first, followed by the init's needed by passed-in args.

jlmucb replied:
OK

gapisback commented on src/cc_helpers.cc:
```diff
@@ -82,16 +82,22 @@ extern string gramine_platform_cert;
+}
+
+certifier::framework::cc_trust_data::cc_trust_data() {
+  cc_trust_data_default_init();
```
Default constructor will now call default initialization routine.

gapisback commented on src/cc_helpers.cc:
```diff
@@ -110,29 +116,6 @@ certifier::framework::cc_trust_data::cc_trust_data(const string& enclave_type, c
-  x509_policy_cert_ = nullptr;
-  cc_is_certified_ = false;
-
-  for (int i = 0; i < num_certified_domains_; i++) {
```
In the old code, this variable `num_certified_domains_` is used uninitialized.
That's causing a seg-fault.

jlmucb replied:
great!

gapisback commented on src/cc_helpers.cc:
```diff
@@ -110,29 +116,6 @@ certifier::framework::cc_trust_data::cc_trust_data(const string& enclave_type, c
-        certified_domains_[i] = nullptr;
-      }
-  }
-  delete certified_domains_;
```
This and delete on L127 seem like destructor code.

To me, seems like this old code was mixing-up logic / code for constructor and destructor.

gapisback commented on tests/pytests/test_cert_framework_basic.py:
```diff
@@ -53,6 +53,29 @@ def test_getmembers_of_cc_framework():
+    """
+    Basic exerciser for an empty cc_trust_data() object.
+    """
+    cctd = libcf.new_cc_trust_data()
```
Without the c++ fix, this single line will trip the seg-fault consistently on every new run.

jlmucb reviewed (changes requested):
Good catch!

The only concern I have is that the destructor does not delete the certifiers structures that were newed when adding.

You are correct that the code that does that, below, should be in the destructor (Man, that was a wierd mistake).

for (int i = 0; i < num_certified_domains_; i++) {
  if (certified_domains_[i] != nullptr) {
    delete certified_domains_[i];
    certified_domains_[i] = nullptr;
  }
}
delete certified_domains_;
certified_domains_ = nullptr;
num_certified_domains_ = 0;
------------------------------------

If you agree and fix that, I'm good with this.  I'm glad you found the problem. Good work.

gapisback commented on src/cc_helpers.cc:
```diff
@@ -117,6 +117,15 @@ void certifier::framework::cc_trust_data::cc_trust_data_default_init() {
 }
 
 certifier::framework::cc_trust_data::~cc_trust_data() {
+  for (int i = 0; i < num_certified_domains_; i++) {
```
@jlmucb -- Restored the correct delete / release-memory code in the desctructor.

See my notes in the test case.

gapisback commented on src/cc_helpers.cc:
```diff
@@ -1247,7 +1256,7 @@ bool certifier::framework::cc_trust_data::certify_secondary_domain(const string&
   }
 
-  return found->certify_domain();
+  return (found ? found->certify_domain() : false);
```
This fixes another seg-fault in this method. If supplied domain-name is not found, `found` will be NULL.

Handle that.

gapisback commented on tests/pytests/test_cert_framework_basic.py:
```diff
@@ -76,6 +76,45 @@ def test_cf_cc_trust_data():
+                                                          'localhost', 8123)
+    assert result is True
+
+    libcf.delete_cc_trust_data(cctd)
```
This delete call in new test case will invoke the `cc_trust_data()` destructor.

I didn't quite find a way to confirm or check thru pytests that there is a memory leak when the new C++ code in the destructor was not added.

But from code inspection, the memory release code in the desctructor is fine, so I'm ok with just this exerciser unit test.

gapisback commented on tests/pytests/test_cert_framework_basic.py:
```diff
@@ -76,6 +76,45 @@ def test_cf_cc_trust_data():
+                                                          'localhost', 8123)
+    assert result is True
+
+    result = libcf.cc_trust_data_certify_secondary_domain(cctd,
```
This lines triggers, and verifies the fix for, the seg-fault in this `certify_secondary_domain()` method.

gapisback commented:
@jlmucb -- I have addressed your comment to add the code in the destructor.

Also, found and fixed another seg-fault. Plus added 2 new tests.

ready for re-review.

I didn't run the multi-domain app exercise workflow ... as it's too time consuming. I think the changes are fine based on inspection.

jlmucb reviewed (approved):
Good catch!

The only concern I have is that the destructor does not delete the certifiers structures that were newed when adding.

You are correct that the code that does that, below, should be in the destructor (Man, that was a wierd mistake).

for (int i = 0; i < num_certified_domains_; i++) {
  if (certified_domains_[i] != nullptr) {
    delete certified_domains_[i];
    certified_domains_[i] = nullptr;
  }
}
delete certified_domains_;
certified_domains_ = nullptr;
num_certified_domains_ = 0;
------------------------------------

If you agree and fix that, I'm good with this.  I'm glad you found the problem. Good work.
